aesd-char-driver/main.c

```diff
@@ -0,0 +1,260 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ * Reference: 1) https://github.com/cu-ecen-aeld/ldd3/blob/master/scull/main.c 
+ * 	      2) Linux Device Drive Edition Chapter 3
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h>   // file_operations
+#include "aesdchar.h"
+#include <linux/slab.h>
+#include <linux/string.h>
+int aesd_major = 0; // use dynamic major
+int aesd_minor = 0;
+
+MODULE_AUTHOR("Ruthvik R Chanda");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+    PDEBUG("open");
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    int value = 0;
+    unsigned long c = 0;
+    int bytes_read = 0;
+    size_t offset;
+    struct aesd_buffer_entry *temp;
+    struct aesd_dev *dev = filp->private_data;;
+    
+    value = mutex_lock_interruptible(&dev->lock);
+    PDEBUG("read %zu bytes with offset %lld", count, *f_pos);
+
+    if (value!=0)
+    {
+	PDEBUG(KERN_ERR "Mutex was not acquired\n");
+	goto handle_error;
+    }
+
+    temp = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->circular_buffer, *f_pos, &offset);
+
+    if(temp==NULL)
+        goto handle_error;
+
+    if ((temp->size - offset) < count) 
+    {
+        *f_pos = *f_pos + (temp->size - offset);
+        bytes_read = temp->size - offset;
+    } 
+    else 
+    {
+        *f_pos = *f_pos + count;
+        bytes_read = count;
+    }
+
+    c = copy_to_user(buf, temp->buffptr+offset, bytes_read);
+    if (c) 
+    {
+	retval = -EFAULT;
+	goto handle_error;
+    }
+
+    retval = bytes_read;
+
+    handle_error:
+            mutex_unlock(&dev->lock);
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos)
+{
+    ssize_t retval = 0; 
+    struct aesd_buffer_entry write_buffer;
+    struct aesd_dev *dev = filp->private_data;
+    int i, send_packet = 0; 
+    int store = 0, temp_size = 0;
+    char *temp_buff;
+    int value = 0;
+    const char *newbuffer;
+
+    PDEBUG("write %zu bytes with offset %lld", count, *f_pos);
+
+    value = mutex_lock_interruptible(&dev->lock);
+    if (value !=0)
+	{
+		PDEBUG(KERN_ERR "Couldn't acquire Mutex\n");
+		return -EFAULT;
+	}
+
+    temp_buff = (char *)kmalloc(count, GFP_KERNEL);
```

Comment:
@Ruthvik-Chanda I think several of the blocks below can be outside the locking interval, see discussion in class today about this.